docs/authentication.md

```diff
@@ -0,0 +1,318 @@
+---
+description: Learn about authentication patterns in Next.js apps and explore a few examples.
+---
+
+# Authentication
+
+Authentication verifies who a user is, while authorization controls what a user can access. Next.js supports multiple patterns for authentication, each designed for different use cases. This guide will allow you to choose your adventure based on your constraints.
+
+## Patterns
+
+The first step to identifying which authentication pattern you need is understanding the data-fetching strategy you want. We can then determine which authentication providers support this strategy. There are two main strategies:
+
+- Use static generation to server-render a loading state, followed by fetching user data client-side.
+- Fetch user data server-side to eliminate a flash of unauthenticated content.
+
+### Static Generation
+
+Next.js automatically determines that a page is static if there are no blocking data requirements. This means the absence of `getServerSideProps` and `getInitialProps` in the page. Instead, your page can render a loading state from the server, followed by fetching the user client-side.
+
+One advantage of this pattern is it allows pages to be served from a global CDN and [preloaded](https://nextjs.org/docs/api-reference/next/link) using `<Link />`. Depending on the size of your application, [React hydration](https://reactjs.org/docs/react-dom.html#hydrate) could take some time. If you serve a loading shell from the server, this allows React to hydrate while you're fetching user data. In practice, this results in a faster TTI ([Time to Interactive](https://web.dev/interactive/)).
+
+Let's look at an example for a profile page. This will initially render a loading skeleton. Once the request for a user has finished, it will show the user's name.
+
+```jsx
+// pages/profile.js
+
+export default Profile = () => {
+  // Fetch the user client-side
+  const { user } = useUser({ redirectTo: '/login' });
+
+  // Server-render loading skeleton
+  if (!user) {
+    return <LoadingSkeleton />;
+  }
+
+  // Once the user request finishes, show the user's name
+  return <p>{user.name}</p>;
+};
+```
```

Comment:
This code samples are very cool and I think they explain very well the idea. Having said that, we should provide a real example here that uses this same code and recommend checking that one if you want to see it on action. This explains the pattern but it's too vague on how an app can have auth.

What I would like to see is something like what we have for the [Environment Variables Documentation](https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser) - There we explain how it works, but also provide a single example that uses every single thing. The example itself has some cool styles and provides more docs too. We could have an example that has multiple routes and shows how auth works in different data fetching scenarios.

Reply:
This was a modified version of the `with-iron-session` example. I will update this to pull directly from there and link back to it.